libs/MotionKit/source/MotionKit.cpp

```diff
@@ -58,56 +56,54 @@ void MotionKit::startStabilisation()
 {
 	stop();
 
-	_imukit.start();
-	_imukit.setOrigin();
+	_previous_euler_angles = _imukit.getEulerAngles();
+	_pid.setTargetYaw(_previous_euler_angles.yaw);
 
 	_target_not_reached		  = false;
 	_stabilisation_requested  = true;
 	_rotate_x_turns_requested = false;
 
-	_event_loop.start();
+	auto on_euler_angles_rdy_callback = [this] { run(); };
+
+	_imukit.onEulerAnglesRdy(on_euler_angles_rdy_callback);
 }
 
 // LCOV_EXCL_START - Dynamic behavior, involving motors and time.
 void MotionKit::run()
 {
-	auto last_yaw			= kReferenceAngle;
-	auto rotations_executed = 0;
+	auto must_rotate = [&] { return _rotate_x_turns_requested && _rotations_executed < _rotations_to_execute; };
+	auto motion_end	 = [&] { return !_stabilisation_requested && !_rotate_x_turns_requested && !_target_not_reached; };
+
+	if (motion_end() && _on_rotation_ended_callback != nullptr) {
+		stop();
+		_on_rotation_ended_callback();
 
-	auto must_rotate = [&] { return _rotate_x_turns_requested && rotations_executed != _rotations_to_execute; };
+		_imukit.onEulerAnglesRdy(nullptr);
+		return;
+	}
 
 	auto check_complete_rotations_executed = [&](auto current_yaw) {
-		if (std::abs(last_yaw - current_yaw) >= 300.F) {
-			++rotations_executed;
+		if (std::abs(_previous_euler_angles.yaw - current_yaw) >= 300.F) {
```

Comment:
are you sure `300.F` is still working?

yaw is now `-180..+180`

you might need to map the values on `0..359`

it was already the case before, but I still can't wrap my head around this way of calculating rotations.

especially now that we don't reset the origin, if you start close to `~350`, it will count rotations for just a few degrees.

The right way to do it would be to accumulate the rotation angle over time and check if you have reached `~359`

it's possible to count clockwise and counter clockwise as well